onnxruntime/core/providers/cuda/nn/instance_norm_impl.cu

```diff
@@ -7,53 +7,54 @@
 namespace onnxruntime {
 namespace cuda {
 
-template <typename T>
+template <typename T1, typename T2>
 __global__ void _InstanceNormKernel(
-    const T* input_data,
-    const T* scale,
-    const T* bias,
-    const T* mean,
-    const T* variance,
+    const T1* __restrict__ input_data,
+    const T1* __restrict__ scale,
+    const T1* __restrict__ bias,
+    const T2* __restrict__ mean,
+    const T2* __restrict__ variance,
     const double variance_correction,
     const double epsilon,
     const fast_divmod fdm_HW,
     const fast_divmod fdm_C,
-    T* output_data,
+    T1* __restrict__ output_data,
     const CUDA_LONG N) {
   CALCULATE_ELEMENTWISE_INDEX_OR_EXIT(id, N);
```

Comment:
>CALCULATE_ELEMENTWISE_INDEX_OR_EXIT(id, N); [](http://example.com/codeflow?start=2&length=43)

I think the issue with many of our kernel performances is that we always use the same block sizes and threads per block. That way, if we the input is too large, we donot have enough threads, some data portions wait for kernels to finish and re-lunch. This describes a better strategy:
[https://developer.nvidia.com/blog/cuda-pro-tip-write-flexible-kernels-grid-stride-loops/](https://developer.nvidia.com/blog/cuda-pro-tip-write-flexible-kernels-grid-stride-loops/)

Reply:
I am aware of grid stride kernels and as you say it *could* be an issue with some kernels when trying to parallelize operations on a large volume of data since we don't clamp number of blocks and handle striding across grids. 
However, this isn't what the PR is about nor the intended scope of this PR.